File: easy/0278_first_bad_version/solution.cpp
using namespace std;

bool isBadVersion(int version);

class Solution {
public:
  int firstBadVersion(int n) {
    unsigned int left = 1;
    unsigned int right = n;

    unsigned int mid;
    bool isBad;

    while (left <= right) {
      mid = (left + right) / 2;

      isBad = isBadVersion(mid);

      if (isBad && !isBadVersion(mid - 1))
        return mid;

      if (!isBad)
        left = mid + 1;
      else
        right = mid - 1;
    }

    return -1;
  }
};

File: easy/0278_first_bad_version/solution.py
def isBadVersion(version: int) -> bool:
    pass


class Solution:
    def firstBadVersion(self, n: int) -> int:
        left = 1
        right = n

        while left <= right:
            mid = (left + right) // 2

            is_bad_version = isBadVersion(mid)

            if is_bad_version and not isBadVersion(mid - 1):
                return mid

            if not is_bad_version:
                left = mid + 1
            else:
                right = mid - 1
